### DISCUSSION TITLE
End to end testing with Directus

### DISCUSSION THREAD

**Discussion by Pachat** [directus/directus#11507] [Archive]
Is there any tool you would recommend?
I tried `Selenium IDE` but it stops in any side page opened by "Add Existing" incapable of selecting and saving the choice.

**Comment by Pachat** [comment]
Well. It works with Cypress.

BUT there is a big hurdle for any tool test: with automatic recording, we need stable identifiers.

For instance, a click on the `+` button generates
`cy.get('[data-tooltip="oFFfa5jq0bnkb4fq08nsx"] > .button > .content > .v-icon > i').click();`

at the next execution, for the same button, it changes to
`cy.get('[data-tooltip="COAT6kO6OzFMfLxinujLT"] > .button > .content > .v-icon > i').click();`

=> These changes are a real challenge when it comes to write stable automatic tests.

**Stable unique identifiers would help a lot**

--

We have the same problem with inputs. For instance, the automatic generation is
`cy.get(':nth-child(2) > .interface > .v-input > .input > input').type('_Something_');`
which would fail, if we change the field's place.

But there, it is easier to circumvent, as we can manually modify the test to
`cy.get('input[field="_fieldname_"]').type('_Something_');`


So, the question is: **how can we generate `id`s for buttons or for any other Directus element?**
If you could point out in which file have a look (where the DOM is build), that may help.

**Comment by rijkvanzanten** [comment]
We haven't seriously looked at adding automated end-to-end tests for the front-end app yet, so I won't be able to comment on it in much detail. 

Stable unique identifiers for every element in Directus sounds like a very difficult task to achieve, without having some sort of automated way to generate those identifiers in a consistent way 🤔 I wouldn't want to have to add and maintain a `id="<uuid>"` on every single element used in Directus by hand

**Reply by Pachat** [reply]
I understand that generation ids may add overwork. So, I changed my mind:
as a matter of fact, I ended using `collection` and `field` that are already here for most of the types [#11817](https://github.com/directus/directus/discussions/11817#discussioncomment-2239340)

However there are still the relationship selectors that do not provide `collection` and `field` 
_It does not need to be the collection to which they refer, 
the collection in which there are is fine: the important thing here is to have stable selectors._

IMHO, as it exists for 90%, it should not be overwork to have `collection` and `field` for relationship selectors.

**Reply by rijkvanzanten** [reply]
I believe those attributes are there as per Vue's default behavior to associate the props that aren't directly used in components as attributes in HTML. Those aren't added manually or on purpose

**Reply by Pachat** [reply]
As said [here](https://github.com/directus/directus/discussions/11817#discussioncomment-2250957) , I did not find them:

With relations fields, like M2O, we don't get the collection, neither the fieldname:

    <div class="interface" data-v-6f93e439="" data-v-c7b47d1e="">
        <div class="many-to-one" autofocus="false" loading="false" width="half" type="integer"
             field-data="[object Object]" primary-key="3" data-v-1b0c186d="" data-v-6f93e439="">...

   
It would be nice to have the collection and field name properties for relations fields.
    That would give more to have more stable debugging scripts

**Reply by rijkvanzanten** [reply]
Yeah I understand. I'm saying that we don't manually add those collection/field attributes on purpose _anywhere_, so the couple that do exist are purely a coincidence

**Reply by Pachat** [reply]
Thank you for your answer.
So the question may raise when you will "look at adding automated end-to-end tests" which are necessary to deliver reliable apps.

**Reply by rijkvanzanten** [reply]
Yup, that's on ongoing process. We're currently focussing on adding the full test coverage for the API first 👍🏻
